### PR TITLE
feat: Staking integration

### PR DESCRIPTION
**Pull request by 0xonramp**
- **Update token mapping**
- **Add staking integration**
- **Add refetch logic**

### CODE REVIEW

File: src/app/prt-staking/constants.ts [hunk 1]
```diff
@@ -4,8 +4,9 @@ import { ProductRevenueToken } from '@/app/prt-staking/types'
 
 export const prts: ProductRevenueToken[] = [
   {
-    tokenData: getIndexTokenData('hyETH')!,
-    prtTokenData: getIndexTokenData('prtHyETH')!,
+    rewardTokenData: getIndexTokenData('hyETH')!,
+    stakeTokenData: getIndexTokenData('prtHyETH')!,
+    stakedTokenData: getIndexTokenData('prtHyETH')!, // FIXME: Add staked token once deployed
```
Comment (0xonramp) [inline src/app/prt-staking/constants.ts:9]: FIXME

File: src/app/prt-staking/components/prt-section/claimable-panel.tsx [hunk 2]
```diff
@@ -8,15 +8,15 @@ export function ClaimablePanel() {
         <div className='flex-1'>Claimable</div>
         <div>
           {claimableRewards
-            ? `${claimableRewards} ${token?.tokenData.symbol}`
+            ? `${claimableRewards} ${token?.rewardTokenData.symbol}`
```
Comment (0xonramp) [inline src/app/prt-staking/components/prt-section/claimable-panel.tsx:11]: hyETH itself

File: src/app/prt-staking/provider.tsx [hunk 3]
```diff
@@ -54,38 +58,69 @@ export const PrtStakingContextProvider = ({ children, token }: Props) => {
   const [cumulativeRevenue, setCumulativeRevenue] = useState<number | null>(
     null,
   )
-  const prtTokenAddress = token.prtTokenData.address as Address
-  const prtTokenDecimals = token.prtTokenData.decimals
+  const stakeTokenAddress = token.stakeTokenData.address as Address
+  const stakedTokenAddress = token.stakedTokenData.address as Address
```
Comment (0xonramp) [inline src/app/prt-staking/provider.tsx:62]: @janndriessen we'll likely need to tweak which one is used in the calls below during testing

File: src/app/prt-staking/provider.tsx [hunk 4]
```diff
@@ -54,38 +58,69 @@ export const PrtStakingContextProvider = ({ children, token }: Props) => {
...
+      query: {
+        enabled: isAddress(address ?? ''),
+      },
+    })
+
+  const { data: canStake } = useReadContract({
+    abi: PrtStakingAbi,
+    address: stakedTokenAddress,
+    functionName: 'canStake',
+  })
```
Comment (0xonramp) [inline src/app/prt-staking/provider.tsx:97]: FIXME: update ABI once `canStake` is added and integrate

File: src/app/prt-staking/provider.tsx [hunk 5]
```diff
@@ -54,38 +58,69 @@ export const PrtStakingContextProvider = ({ children, token }: Props) => {
...
+
+  const { data: isApprovedStaker } = useReadContract({
     abi: PrtStakingAbi,
-    address: prtTokenAddress,
-    functionName: 'getPendingRewards',
+    address: stakedTokenAddress,
+    functionName: 'isApprovedStaker',
```
Comment (0xonramp) [inline src/app/prt-staking/provider.tsx:102]: Confirm this is the correct address used for this function

File: src/app/prt-staking/provider.tsx [hunk 6]
```diff
@@ -96,68 +131,103 @@ export const PrtStakingContextProvider = ({ children, token }: Props) => {
...
+          address: stakeTokenAddress,
+          functionName: 'stake',
+          args: [amount],
+        })
+      } else {
+        if (!stakeSignMessage) return
+        if (!stakeDomain) return
+        const signature = await walletClient.signTypedData({
+          // FIXME: Domain arg typing/format
+          domain: stakeDomain,
```
Comment (0xonramp) [inline src/app/prt-staking/provider.tsx:169]: Domain value returned from the contract doesn't match the expected type here, will need to investigate

Comment (janndriessen) [reply src/app/prt-staking/provider.tsx:169]: Usually this is a struct (https://eips.ethereum.org/EIPS/eip-712#definition-of-domainseparator) with at least `version`and `verifyingContract` defined.

Comment (janndriessen) [reply src/app/prt-staking/provider.tsx:169]: As mentioned on discord, I would just hard-code this. No need to fetch every time.

File: src/app/prt-staking/components/prt-section/prt-widget.tsx [hunk 7]
```diff
@@ -51,28 +63,39 @@ export function PrtWidget({ token, onClose }: Props) {
     return ''
   }, [currentTab])
 
-  const onClickTradeButton = useCallback(() => {
+  const onClickTradeButton = useCallback(async () => {
     if (currentTab === WidgetTab.STAKE) {
-      stakePrts(parseUnits(inputAmount, token.prtTokenData.decimals))
+      if (!isApproved) {
+        await onApprove()
```
Comment (0xonramp) [inline src/app/prt-staking/components/prt-section/prt-widget.tsx:69]: Ensure the user has approved the token

File: src/app/prt-staking/components/prt-section/prt-widget.tsx [hunk 8]
```diff
@@ -101,14 +124,15 @@ export function PrtWidget({ token, onClose }: Props) {
...
+        showSelectorButton={currentTab !== WidgetTab.CLAIM}
         showSelectorButtonChevron={false}
         onChangeInput={(_, amount) => setInputAmount(amount)}
         onClickBalance={onClickBalance}
         onSelectToken={() => {}}
       />
       <TradeButton
         label={buttonLabel}
-        isDisabled={false}
+        isDisabled={currentTab === WidgetTab.STAKE && !canStake}
```
Comment (0xonramp) [inline src/app/prt-staking/components/prt-section/prt-widget.tsx:135]: Leverage the new `canStake` functionality (ensure the user does not stake during the buffer time)

File: src/app/prt-staking/components/prt-section/prt-widget.tsx (hunk 7 above)
Comment (janndriessen) [reply src/app/prt-staking/components/prt-section/prt-widget.tsx:69]: Should there be a `return`?

Comment (0xonramp) [reply src/app/prt-staking/components/prt-section/prt-widget.tsx:69]: @janndriessen not in this case - we want to continue with the other operations (stake and then refetch)

File: src/lib/utils/index.ts [hunk 9]
```diff
@@ -13,6 +16,26 @@ export function isSameAddress(address1: string, address2: string): boolean {
   return address1.toLowerCase() === address2.toLowerCase()
 }
 
+// TODO: Settle on one token typing approach?
```
Comment (janndriessen) [inline src/lib/utils/index.ts:19]: Personally, I think it's ok that the app has its own token type - as we're adding a lot of extra properties that wouldn't be useful in tokenlists. I think it's common to have a kind of adapter pattern for these things. But if annoying we could look into maybe handling the extra properties in different ways. I think lots of them are kind of obsolete now.

**Review by janndriessen — APPROVED** [review]

**Pull request merged**